accepted/rfc0000-scope-scheduling.md

```diff
@@ -0,0 +1,334 @@
+there are some advantages to offering more scheduling modes:
+
+[stjepang-1]: https://github.com/rayon-rs/rfcs/pull/1#issuecomment-437074748
+
+- Per-thread LIFO is the **current default behavior** of scopes. This
+  is "semi-documented". While not strictly part of our semver
+  guarantees, altering this behavior could negatively affect existing
+  applications.
+- Per-thread LIFO offers the **most efficient** implementation in a
+  "micro" sense, as it can build directly on the work-stealing deques
+  and does not require any indirection. It also has desirable cache
+  behavior, as threasd will tend to continue the "most recent thing"
+  you were doing, which is also the thing where the caches are
+  warmest, etc. **These two cases together argue that, for cases where
+  the application doesn't otherwise care about execution order,
+  per-thread LIFO is a good choice.** This also applies to
+  applications that wish to choose another scheduling constraint (see
+  the next section).
+- Global FIFO, meanwhile, is the only scheduling order of the 3 that is
+  independent from the work-stealing behavior.
+
+**Should we offer additional scheduling modes?** Another question
+worth considering is "why stop with these three modes"?  For example,
+the [Rayon demo application for solving the Travelling Salesman
+Problem][tsp] also [uses the "indirect task"
+trick][tsp-indirect]. However, in that case, it uses a (locked)
+priority queue to control the ordering. Rayon could conceivably offer
+some more flexible, priority-queue like mechanism as well. However,
+it's not clear that this is worth doing, since one can always achieve
+the same effect in "user space", as the TSP application does. (For
+this purpose, per-thread LIFO tasks are a great fit, as they add
+minimal overhead.)
+
+[tsp]: https://github.com/rayon-rs/rayon/blob/a68b05ce524f79d7e7a5065714a8d3ca40ce8d4b/rayon-demo/src/tsp/
+[tsp-indirect]: https://github.com/rayon-rs/rayon/blob/a68b05ce524f79d7e7a5065714a8d3ca40ce8d4b/rayon-demo/src/tsp/step.rs#L50-L51
+
+**Is a builder really necessary?** We could simply add some more
+variants of `rayon::scope`, such as
+`rayon::scope_per_thread_fifo`. Perhaps that would be easier. We can
+always move to a builder at some later point.
+
+**Should we encode the scheduling mode in the type of the scope?** The
+prototype implementation of `Scope` stores the scheduling mode as an
+"enum". An alternative would be to use generics to encode the mode;
+this would avoid a dynamic `if` to select between per-thread
+LIFO/per-thread FIFO etc, but at the cost of more complexity in the
+user-visible types (it would also mean that we must expose more than
+one scope type, as the existing -- stable -- rayon-core API does not
+contain a generic). Given the performance measurements, it is unlikely
+that the more complex types are worthwhile. This change could also be
+made at some later time, conceivably, though it would require more
+additions to the `ScopeBuilder` API.
```

Comment:
How could it be changed later?  If we expose per-thread FIFO using the current `Scope` type now, it seems to me that we'll be locked into that.

Reply:
I think I meant "if we use a distinct type, we can still use a branch in the implementation" -- I'll tweak the wording